### PR TITLE
Hide defaulted optional values from Flow UI 

### LINKED ISSUE #2811: Flow API: Show only some non-nil values in the UI
A bugfix in #2800 caused all non-nil attributes to be shown in the Flow UI. This was deliberate, in order to fix #2206 and #2205. However, the information which is displayed after this bugfix may be too much, especially for components with a lot of attributes. A few ideas were discussed:

- @rfratto suggested only showing the attributes which were explicitly set in the Agent's config + all the non-zero attributes for a given component. 

- @erikbaranowski suggested a River tag to explicitly hide/show things in the config. This could be useful when there is a default value in an optional attribute which could be significant.

- @mattdurham suggested that there is a benefit in showing anything non-nil, because even optional fields with default values could be important.

A few other considerations were brought up:

- @rfratto mentioned that nil never shows up in River except for capsules.

- @rfratto mentioned that at at some point we need to trim down the config because in the future it needs to be stored on Grafana Cloud efficiently.

- @tpaschalis mentioned that good example components to look into are `prometheus.remote_write` and any `otelcol` ones which have an http client, since those would have lots of optional attributes.

For now the consensus seems to be to go with Robert's suggestion to show everything configured by the user + any non-zero attributes. 

By "non-zero" we mean attributes which are not the zero value of their type. Examples of zero values would be a boolean which is "false", an integer which is "0" or a string which is "".

### PR DESCRIPTION
**Pull request by rfratto**
This PR extends the work @erikbaranowski has done in #3998 and #4057 to the UI, where attributes set to their default values will be hidden from the UI. Default values are either retrieved via the `river.Defaulter` implementation, or by the respective Go zero values if a river.Defaulter implementation doesn't exist for a type. 

To implement this work more easily, I refactored the `pkg/river/encoding` package as `pkg/river/encoding/riverjson`, with the following benefits:

* The package name is now more clear about what it does, rather than generic "encoding".
* The implementation aligns more closely with pkg/river/token/builder, which allows fixes to be ported to both more easily. 
* The implementation is about half the size as it was previously, ideally making it easier to maintain long-term.

Closes #2811, as it addresses the problem where an attribute set to a zero value was hidden from the UI, even if that zero value was not the default. 

I used cmd/grafana-agent/example-config.river with this change, and the UI appears to still work fully as expected.

### CODE REVIEW

File: pkg/river/encoding/riverjson/types.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package riverjson
+
+// Various concrete types used to marshal River values.
+type (
+	// jsonStatement is a statement within a River body.
+	jsonStatement interface{ isStatement() }
+
+	// A jsonBody is a collection of statements.
```
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/river/encoding/riverjson/types.go:8]: I can make this change but JSYK standard godoc comments permit leading articles priort to the identifier like A/An/The.

[the remark this answers is not in the document]
Comment (erikbaranowski) [reply pkg/river/encoding/riverjson/types.go:8]: nvm then 👍

**Review by erikbaranowski — APPROVED** [review]
Looks great! A big improvement to the package layout plus the new default/zero checking logic.

I validated locally with my own config.

**Review by erikbaranowski — APPROVED** [review]

**Review by mattdurham — APPROVED** [review]
Clean!

**Pull request merged**